### PR TITLE
Exercise 7 adding redux

### PR DESCRIPTION
**Pull request by sashmen5**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -40,5 +53,25 @@
...
+    "lint-staged": "^8.1.6",
+    "prettier": "^1.17.1",
+    "pretty-quick": "^1.10.0"
+  },
+  "prettier": {
+    "singleQuote": true
+  },
+  "lint-staged": {
+    "*.{ts,tsx}": [
+      "prettier --write --single-quote=true",
```
Comment (sashmen5) [inline package.json]: delete --single

File: src/App.tsx [hunk 2]
```diff
@@ -19,27 +21,35 @@ const GlobalStyle = createGlobalStyle`
...
-        </Router>
-    );
+  return (
+    <Router>
+      <ThemeProvider theme={mainTheme}>
+        <Content>
+          <GlobalStyle />
+          <Switch>
+            <Route path="/" exact component={Login} />
+            <Route path={`/${PROJECTS}`} exact component={Projects} />
```
Comment (ronlanger) [inline src/App.tsx]:
```suggestion
            <Route path={PROJECTS} exact component={Projects} />
```

File: src/components/Login.tsx [hunk 3]
```diff
@@ -18,47 +23,47 @@ const LoginTitle = styled(Title)`
...
+      setIsLoading(false);
+      setToHomePage(true);
+    }, 1000);
+  };
 
-    if (toHomePage) {
-        return <Redirect push to='/projects' />
-    }
+  if (toHomePage) {
+    return <Redirect push to="/projects" />;
```
Comment (ronlanger) [inline src/components/Login.tsx]: const

File: src/components/ProjectItem.tsx [hunk 4]
```diff
@@ -24,28 +30,38 @@ const DangerousLabel = styled(Label)`
...
+const ProjectItem: FC<ProjectItemProps> = ({
+  project,
+  handleDeleteProject,
+  handleEditProject,
+  itemsDirection
+}: ProjectItemProps) => {
+  const { id } = project;
+  return (
+    <Container flexDirection={itemsDirection}>
+      <StyledLink to={{ pathname: `${PROJECT}/${project.id}` }}>
```
Comment (ronlanger) [inline src/components/ProjectItem.tsx]: same

File: src/components/ProjectTracker.tsx [hunk 5]
```diff
@@ -23,47 +34,52 @@ const ProjectLabel = styled.div`
...
+      />
+      <TimeTrackingContainer>
+        <ProjectLabel>{projectLabel}</ProjectLabel>
+        {dateLabels.map(item => (
+          <TimeSlot
+            key={item}
+            dateLabel={item}
+            projectId={project}
+            amountOfHours={
+              timeSlots[item] && timeSlots[item].amountOfHours
```
Comment (ronlanger) [inline src/components/ProjectTracker.tsx]: extract to const and maybe use lodash/fp

File: src/components/Projects.tsx [hunk 6]
```diff
@@ -24,11 +37,11 @@ const ModalContent = styled.div`
 `;
 
 interface ModalWrapperProps {
-    openModal: boolean
+  openModal: boolean;
 }
 
 const ModalWrapper = styled.div<ModalWrapperProps>`
-  visibility: ${props => props.openModal ? '' : 'hidden'};
+  visibility: ${props => (props.openModal ? '' : 'hidden')};
```
Comment (ronlanger) [inline src/components/Projects.tsx]:
```suggestion
  ${props => (props.openModal && 'visibility : hidden')};
```

File: src/components/TimeSlot.tsx [hunk 7]
```diff
@@ -1,55 +1,64 @@
...
-                onChange={e => handleInputChange(e)}
-            />
-        </Container>
-    )
+  return (
+    <Container>
+      <div>{dateLabel}</div>
+      <TimeSlotInput
+        type="text"
+        value={amountOfHours ? amountOfHours : ''}
```
Comment (ronlanger) [inline src/components/TimeSlot.tsx]:
```suggestion
        value={amountOfHours || ''}
```

File: src/store/reducers/projectsReducer.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  2: { id: 2, label: 'Wix' },
+  3: { id: 3, label: 'Facebook' },
+  4: { id: 4, label: 'Apple' }
+};
+
+const projectsReducer: Reducer<Keyed<Project>> = handleActions(
+  {
+    [ADD_PROJECT]: (state: Keyed<Project>, action: AddProjectAction) => {
+      const { id, label } = action.payload;
+      return set(`${id}`, { id, label }, state);
```
Comment (ronlanger) [inline src/store/reducers/projectsReducer.ts]:
```suggestion
      return set(id, { id, label }, state);
```

File: src/store/reducers/timeTrackerReducer.ts [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+  projectsTimeSlots: {},
+  startDate: new Date()
+};
+
+const timeTrackerReducer: Reducer<TimeTracker> = handleActions(
+  {
+    [EDIT_TIME_TRACKER]: (state: TimeTracker, action: EditTimeTracker) => {
+      const { projectId, date, amountOfHours } = action.payload;
+      return set(
+        `projectsTimeSlots.${projectId}[${date}]`,
```
Comment (ronlanger) [inline src/store/reducers/timeTrackerReducer.ts]:
```suggestion
        ['projectsTimeSlots', projectId, date]`,
```

File: src/components/ProjectTracker.tsx (hunk 5 above)
Comment (sashmen5) [reply src/components/ProjectTracker.tsx]: extract to func

File: src/components/TimeSlot.tsx (hunk 7 above)
Comment (sashmen5) [reply src/components/TimeSlot.tsx]: done

File: src/store/reducers/projectsReducer.ts (hunk 8 above)
Comment (sashmen5) [reply src/store/reducers/projectsReducer.ts]: fixed

**Pull request merged**